File: leetcode/1822. Sign of the Product of an Array/main.ex
defmodule Solution do
  @moduledoc """
    Uses list comprehension to reduce the input into the product of each component of `nums`, and then retrieves its sign.
    We do not need to first map each component to its sign because Elixir does not have a limit on the integer size.
  """
  
  @spec array_sign(nums :: [integer]) :: integer
  def array_sign(nums) do
    for x <- nums, reduce: 1 do
        acc ->
            x * acc
    end
    |> sign_func()
  end

  def sign_func(x) when x < 0, do: -1
  def sign_func(x) when x > 0, do: 1
  def sign_func(0), do: 0
end

File: leetcode/1822. Sign of the Product of an Array/main.rs
impl Solution {
    pub fn array_sign(nums: Vec<i32>) -> i32 {
        // need to call sign_func on each component of the array
        // because otherwise we can overflow i32 (or even i64)
        // since the product can be 100^1000 for given constraints
        (nums.iter().fold(1, |acc, &b| acc * Solution::sign_func(b)))
    }

    fn sign_func(x: i32) -> i32 {
        match x.cmp(&0) {
            std::cmp::Ordering::Less => -1,
            std::cmp::Ordering::Equal => 0,
            std::cmp::Ordering::Greater => 1,
        }
    }
}